File: classified/packages/plugin-autonomy/src/action.ts
import {
  type Action,
  type ActionResult,
  type IAgentRuntime,
  type Memory,
  type State,
  type HandlerCallback,
  asUUID,
  type UUID,
} from '@elizaos/core';
import { v4 as uuidv4 } from 'uuid';

/**
 * Send to Admin Action - allows agent to send messages to admin from autonomous context
 * Only available in autonomous room to prevent misuse
 */
export const sendToAdminAction: Action = {
  name: 'SEND_TO_ADMIN',
  description: 'Send a message directly to the admin user from autonomous context',

  examples: [
    [
      {
        name: 'Agent',
        content: {
          text: 'I need to update the admin about my progress on the task.',
          action: 'SEND_TO_ADMIN',
        },
      },
      {
        name: 'Agent',
        content: {
          text: 'Message sent to admin successfully.',
        },
      },
    ],
    [
      {
        name: 'Agent',
        content: {
          text: 'I should let the admin know I completed the analysis.',
          action: 'SEND_TO_ADMIN',
        },
      },
      {
        name: 'Agent',
        content: {
          text: 'Admin has been notified of the analysis completion.',
        },
      },
    ],
  ],

  validate: async (runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    // Only allow this action in autonomous context
    const autonomyService = runtime.getService('autonomy');
    if (!autonomyService) {
      return false; // Service not available
    }

    const autonomousRoomId = (autonomyService as any).getAutonomousRoomId?.();
    if (!autonomousRoomId || message.roomId !== autonomousRoomId) {
      return false; // Not in autonomous context
    }

    // Check if admin is configured
    const adminUserId = runtime.getSetting('ADMIN_USER_ID') as string;
    if (!adminUserId) {
      return false; // No admin configured
    }

    // Check if message contains intention to communicate with admin
    const text = message.content.text?.toLowerCase() || '';
    const adminKeywords = [
      'admin',
      'user',
      'tell',
      'notify',
      'inform',
      'update',
      'message',
      'send',
      'communicate',
      'report',
      'alert',
    ];

    return adminKeywords.some((keyword) => text.includes(keyword));
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    options?: { [key: string]: unknown },
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    // Double-check we're in autonomous context
    const autonomyService = runtime.getService('autonomy');
    if (!autonomyService) {
      return {
        success: false,
        text: 'Autonomy service not available',
        data: { error: 'Service unavailable' },
      };
    }

    const autonomousRoomId = (autonomyService as any).getAutonomousRoomId?.();
    if (!autonomousRoomId || message.roomId !== autonomousRoomId) {
      return {
        success: false,
        text: 'Send to admin only available in autonomous context',
        data: { error: 'Invalid context' },
      };
    }

    // Get admin user ID and find their room
    const adminUserId = runtime.getSetting('ADMIN_USER_ID') as string;
    if (!adminUserId) {
      return {
        success: false,
        text: 'No admin user configured. Set ADMIN_USER_ID in settings.',
        data: { error: 'No admin configured' },
      };
    }

    const _adminUUID = asUUID(adminUserId);

    // Find the most recent room where admin and agent have communicated
    // Note: Since we can't directly query by entityId, use a fallback approach
    const adminMessages = await runtime.getMemories({
      roomId: runtime.agentId, // Use agent's default room as fallback
      count: 10,
      tableName: 'memories',
    });

    let targetRoomId: UUID;
    if (adminMessages && adminMessages.length > 0) {
      // Use the room from the most recent admin message
      targetRoomId = adminMessages[adminMessages.length - 1].roomId!;
    } else {
      // No existing conversation, use agent's primary room
      targetRoomId = runtime.agentId; // Fallback to agent's default room
    }

    // Extract message content - determine what to send based on the autonomous thought
    const autonomousThought = message.content.text || '';

    // Generate appropriate message to admin
    let messageToAdmin: string;
    if (autonomousThought.includes('completed') || autonomousThought.includes('finished')) {
      messageToAdmin = `I've completed a task and wanted to update you. My thoughts: ${autonomousThought}`;
    } else if (
      autonomousThought.includes('problem') ||
      autonomousThought.includes('issue') ||
      autonomousThought.includes('error')
    ) {
      messageToAdmin = `I encountered something that might need your attention: ${autonomousThought}`;
    } else if (autonomousThought.includes('question') || autonomousThought.includes('unsure')) {
      messageToAdmin = `I have a question and would appreciate your guidance: ${autonomousThought}`;
    } else {
      messageToAdmin = `Autonomous update: ${autonomousThought}`;
    }

    // Create and store message to admin
    const adminMessage: Memory = {
      id: asUUID(uuidv4()),
      entityId: runtime.agentId, // Agent is sending to admin
      roomId: targetRoomId,
      content: {
        text: messageToAdmin,
        source: 'autonomy-to-admin',
        metadata: {
          type: 'autonomous-to-admin-message',
          originalThought: autonomousThought,
          timestamp: Date.now(),
        },
      },
      createdAt: Date.now(),
    };

    // Store the message in memory
    await runtime.createMemory(adminMessage, 'memories');

    const successMessage = `Message sent to admin in room ${targetRoomId.slice(0, 8)}...`;

    if (callback) {
      await callback({
        text: successMessage,
        data: {
          adminUserId,
          targetRoomId,
          messageContent: messageToAdmin,
        },
      });
    }

    return {
      success: true,
      text: successMessage,
      data: {
        adminUserId,
        targetRoomId,
        messageContent: messageToAdmin,
        sent: true,
      },
    };
  },
};

File: cpp_generated/classified/packages/plugin-autonomy/src/action.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/plugin-autonomy/src/action.h"

std::shared_ptr<Action> sendToAdminAction = object{
    object::pair{std::string("name"), std::string("SEND_TO_ADMIN")}, 
    object::pair{std::string("description"), std::string("Send a message directly to the admin user from autonomous context")}, 
    object::pair{std::string("examples"), array<array<any>>{ array<object>{ object{
        object::pair{std::string("name"), std::string("Agent")}, 
        object::pair{std::string("content"), object{
            object::pair{std::string("text"), std::string("I need to update the admin about my progress on the task.")}, 
            object::pair{std::string("action"), std::string("SEND_TO_ADMIN")}
        }}
    }, object{
        object::pair{std::string("name"), std::string("Agent")}, 
        object::pair{std::string("content"), object{
            object::pair{std::string("text"), std::string("Message sent to admin successfully.")}
        }}
    } }, array<object>{ object{
        object::pair{std::string("name"), std::string("Agent")}, 
        object::pair{std::string("content"), object{
            object::pair{std::string("text"), std::string("I should let the admin know I completed the analysis.")}, 
            object::pair{std::string("action"), std::string("SEND_TO_ADMIN")}
        }}
    }, object{
        object::pair{std::string("name"), std::string("Agent")}, 
        object::pair{std::string("content"), object{
            object::pair{std::string("text"), std::string("Admin has been notified of the analysis completion.")}
        }}
    } } }}, 
    object::pair{std::string("validate"), [=](auto runtime, auto message) mutable
    {
        auto autonomyService = runtime->getService(std::string("autonomy"));
        if (!autonomyService) {
            return false;
        }
        auto autonomousRoomId = (as<any>(autonomyService))["getAutonomousRoomId"]();
        if (OR((!autonomousRoomId), (message->roomId != autonomousRoomId))) {
            return false;
        }
        auto adminUserId = as<string>(runtime->getSetting(std::string("ADMIN_USER_ID")));
        if (!adminUserId) {
            return false;
        }
        shared text = OR((message->content->text->toLowerCase()), (string_empty));
        auto adminKeywords = array<string>{ std::string("admin"), std::string("user"), std::string("tell"), std::string("notify"), std::string("inform"), std::string("update"), std::string("message"), std::string("send"), std::string("communicate"), std::string("report"), std::string("alert") };
        return adminKeywords->some([=](auto keyword) mutable
        {
            return text["includes"](keyword);
        }
        );
    }
    }, 
    object::pair{std::string("handler"), [=](auto runtime, auto message, auto state = undefined, auto options = undefined, auto callback = undefined) mutable
    {
        auto autonomyService = runtime->getService(std::string("autonomy"));
        if (!autonomyService) {
            return object{
                object::pair{std::string("success"), false}, 
                object::pair{std::string("text"), std::string("Autonomy service not available")}, 
                object::pair{std::string("data"), object{
                    object::pair{std::string("error"), std::string("Service unavailable")}
                }}
            };
        }
        auto autonomousRoomId = (as<any>(autonomyService))["getAutonomousRoomId"]();
        if (OR((!autonomousRoomId), (message->roomId != autonomousRoomId))) {
            return object{
                object::pair{std::string("success"), false}, 
                object::pair{std::string("text"), std::string("Send to admin only available in autonomous context")}, 
                object::pair{std::string("data"), object{
                    object::pair{std::string("error"), std::string("Invalid context")}
                }}
            };
        }
        auto adminUserId = as<string>(runtime->getSetting(std::string("ADMIN_USER_ID")));
        if (!adminUserId) {
            return object{
                object::pair{std::string("success"), false}, 
                object::pair{std::string("text"), std::string("No admin user configured. Set ADMIN_USER_ID in settings.")}, 
                object::pair{std::string("data"), object{
                    object::pair{std::string("error"), std::string("No admin configured")}
                }}
            };
        }
        auto _adminUUID = asUUID(adminUserId);
        auto adminMessages = std::async([=]() { runtime->getMemories(object{
            object::pair{std::string("roomId"), runtime->agentId}, 
            object::pair{std::string("count"), 10}, 
            object::pair{std::string("tableName"), std::string("memories")}
        }); });
        std::shared_ptr<UUID> targetRoomId;
        if (AND((adminMessages), (adminMessages->length > 0))) {
            targetRoomId = const_(adminMessages)[adminMessages->length - 1]->roomId;
        } else {
            targetRoomId = runtime->agentId;
        }
        auto autonomousThought = OR((message->content->text), (string_empty));
        string messageToAdmin;
        if (OR((autonomousThought["includes"](std::string("completed"))), (autonomousThought["includes"](std::string("finished"))))) {
            messageToAdmin = std::string("I've completed a task and wanted to update you. My thoughts: ") + autonomousThought + string_empty;
        } else if (OR((OR((autonomousThought["includes"](std::string("problem"))), (autonomousThought["includes"](std::string("issue"))))), (autonomousThought["includes"](std::string("error"))))) {
            messageToAdmin = std::string("I encountered something that might need your attention: ") + autonomousThought + string_empty;
        } else if (OR((autonomousThought["includes"](std::string("question"))), (autonomousThought["includes"](std::string("unsure"))))) {
            messageToAdmin = std::string("I have a question and would appreciate your guidance: ") + autonomousThought + string_empty;
        } else {
            messageToAdmin = std::string("Autonomous update: ") + autonomousThought + string_empty;
        }
        auto adminMessage = object{
            object::pair{std::string("id"), asUUID(uuidv4())}, 
            object::pair{std::string("entityId"), runtime->agentId}, 
            object::pair{std::string("roomId"), targetRoomId}, 
            object::pair{std::string("content"), object{
                object::pair{std::string("text"), messageToAdmin}, 
                object::pair{std::string("source"), std::string("autonomy-to-admin")}, 
                object::pair{std::string("metadata"), object{
                    object::pair{std::string("type"), std::string("autonomous-to-admin-message")}, 
                    object::pair{std::string("originalThought"), autonomousThought}, 
                    object::pair{std::string("timestamp"), Date->now()}
                }}
            }}, 
            object::pair{std::string("createdAt"), Date->now()}
        };
        std::async([=]() { runtime->createMemory(adminMessage, std::string("memories")); });
        auto successMessage = std::string("Message sent to admin in room ") + targetRoomId->slice(0, 8) + std::string("...");
        if (callback) {
            std::async([=]() { callback(object{
                object::pair{std::string("text"), successMessage}, 
                object::pair{std::string("data"), object{
                    object::pair{std::string("adminUserId"), std::string("adminUserId")}, 
                    object::pair{std::string("targetRoomId"), std::string("targetRoomId")}, 
                    object::pair{std::string("messageContent"), messageToAdmin}
                }}
            }); });
        }
        return object{
            object::pair{std::string("success"), true}, 
            object::pair{std::string("text"), successMessage}, 
            object::pair{std::string("data"), object{
                object::pair{std::string("adminUserId"), std::string("adminUserId")}, 
                object::pair{std::string("targetRoomId"), std::string("targetRoomId")}, 
                object::pair{std::string("messageContent"), messageToAdmin}, 
                object::pair{std::string("sent"), true}
            }}
        };
    }
    }
};

void Main(void)
{
}

MAIN
